File: Solutions/PE014_longest_collatz_sequence/cpp/LongestCollatzSequence.cpp
#include <iostream>
#include <map>
#include <vector>

// https://www.hackerrank.com/contests/projecteuler/challenges/euler014

using namespace std;

long nextCollatz(long n) {
	//returns next number in collatz sequence
	if (n % 2 == 0)
		n >>= 1;
	else
		n = 3 * n + 1;
	return n;
}

map<long, long> cache;
const long cachesize = 100000;

long collatzSequenceLength(long n) {
	if (n == 1)
		return 1;
	else {
		if (n < cachesize and cache.find(n) != cache.end())
			return cache[n];
		long m = nextCollatz(n);
		long res = 1 + collatzSequenceLength(m);
		if (n < cachesize)
			cache[n] = res;
		return res;
	}
}

vector<long> longestCollatzSoFar(long N) {
	//returns a vector where index i is the number less than or equal to i that produces the longest collatz sequence
	cache[1] = 1;
	vector<long> longestSoFar;
	for (long i = 0; i < N + 1; i++)
		longestSoFar.push_back(1);
	long i = 2;
	long m = 1;
	long m_index = 1;
	while (i <= N) {
		long size = collatzSequenceLength(i);
		if (size >= m) {
			m = size;
			m_index = i;
		}
		longestSoFar[i] = m_index;
		i++;
	}
	return longestSoFar;
}


int main() {
	vector<long> res = longestCollatzSoFar(5000000);
	long t;
	cin >> t;
	for (long i = 0; i < t; i++) {
		long n;
		cin >> n;
		cout << res[n] << endl;
	}

}

File: Solutions/PE014_longest_collatz_sequence/longest_collatz_sequence.py
#!/bin/python3

# https://www.hackerrank.com/contests/projecteuler/challenges/euler014


def next_collatz(n):
    """
    input is an integer
    returns next number in collatz sequence
    """
    if n % 2 == 0:
        n >>= 1
    else:
        n = 3*n + 1
    return n


cachesize = 5*10**6
cache = [0]*cachesize
cache[1] = 1


def collatz_sequence_length(n):
    if n == 1:
        return 1
    else:
        global cachesize
        global cache
        if n < cachesize and cache[n] != 0:
            return cache[n]
        m = next_collatz(n)
        res = 1 + collatz_sequence_length(m)
        if n < cachesize:
            cache[n] = res
        return res


def longest_collatz_sequence_so_far(N):
    """
    returns a list where index i is the number less than or equal to i that produces the longest collatz sequence
    """
    longest_so_far = [1]*(N+1)
    i = 2
    m = 1
    m_index = 1
    while i <= N:
        size = collatz_sequence_length(i)
        if size >= m:
            m = size
            m_index = i
        longest_so_far[i] = m_index
        i += 1
    return longest_so_far


if __name__ == '__main__':
    res = longest_collatz_sequence_so_far(5000000)
    t = int(input())
    for _ in range(t):
        n = int(input())
        print(res[n])
